plugins/net_plugin/net_plugin.cpp

```diff
@@ -3492,57 +3491,37 @@ namespace eosio {
-                  if (net_version < proto_dup_goaway_resolution || msg.network_version < proto_dup_goaway_resolution) {
-                     // It's possible that both peers could arrive here at relatively the same time, so
-                     // we need to avoid the case where they would both tell a different connection to go away.
-                     // Using the sum of the initial handshake times of the two connections, we will
-                     // arbitrarily (but consistently between the two peers) keep one of them.
-
-                     auto check_time = check->last_handshake_sent.time + check->last_handshake_recv.time;
-                     g_check_conn.unlock();
-                     if (msg.time + c_time <= check_time)
-                        return false;
-                  } else if (net_version < proto_dup_node_id_goaway || msg.network_version < proto_dup_node_id_goaway) {
-                     if (listen_address < msg.p2p_address) {
-                        fc_dlog( logger, "listen_address '${lhs}' < msg.p2p_address '${rhs}'",
-                                 ("lhs", listen_address)( "rhs", msg.p2p_address ) );
-                        // only the connection from lower p2p_address to higher p2p_address will be considered as a duplicate,
-                        // so there is no chance for both connections to be closed
-                        return false;
-                     }
-                  } else if (my_impl->node_id < msg.node_id) {
-                     fc_dlog( logger, "not duplicate, my_impl->node_id '${lhs}' < msg.node_id '${rhs}'",
-                              ("lhs", my_impl->node_id)("rhs", msg.node_id) );
-                     // only the connection from lower node_id to higher node_id will be considered as a duplicate,
-                     // so there is no chance for both connections to be closed
-                     return false;
-                  }
-                  return true;
                }
-               return false;
-            };
-            if (my_impl->connections.any_of_connections(std::move(is_duplicate))) {
-               peer_dlog( this, "sending go_away duplicate, msg.p2p_address: ${add}", ("add", msg.p2p_address) );
-               go_away_message gam(duplicate);
-               gam.node_id = conn_node_id;
-               enqueue(gam);
-               no_retry = duplicate;
-               return;
+               return true;
             }
-         } else {
-            peer_dlog(this, "skipping duplicate check, addr == ${pa}, id = ${ni}", ("pa", peer_address())("ni", msg.node_id));
-
-            // check if peer requests no trx or no blocks
-            set_peer_connection_type(msg.p2p_address);
+            return false;
+         };
+         if (my_impl->connections.any_of_connections(is_duplicate)) {
+            peer_dlog( this, "sending go_away duplicate, msg.p2p_address: ${add}", ("add", msg.p2p_address) );
+            go_away_message gam(duplicate);
```

Comment:
I don't think the constructor of `go_away_message` serves any useful purpose, and makes the code here at least more awkward. Without it we could do:
`enqueue(go_away_message{duplicate, conn_node_id});`

Reply:
bonus point if making `go_away_reason` an `enum class` :-)

Reply:
Done.